src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs

```diff
@@ -87,31 +87,65 @@ private static EnumInfo<TStorage> GetEnumInfo<TStorage>(RuntimeType enumType, bo
                 typeof(TStorage) == typeof(nuint) || typeof(TStorage) == typeof(float) || typeof(TStorage) == typeof(double) || typeof(TStorage) == typeof(char),
                 $"Unexpected {nameof(TStorage)} == {typeof(TStorage)}");
 
-            return enumType.GenericCache is EnumInfo<TStorage> info && (!getNames || info.Names is not null) ?
+            return enumType.FindCacheEntry<EnumInfo<TStorage>>() is {} info && (!getNames || info.Names is not null) ?
                 info :
                 InitializeEnumInfo(enumType, getNames);
 
             [MethodImpl(MethodImplOptions.NoInlining)]
             static EnumInfo<TStorage> InitializeEnumInfo(RuntimeType enumType, bool getNames)
+            {
+                // If we're asked to get the cache with names,
+                // force that copy into the cache even if we already have a cache entry without names
+                // so we don't have to recompute the names if asked again.
+                return getNames
+                    ? enumType.ReplaceCacheEntry(EnumInfo<TStorage>.Create(enumType, getNames: true))
+                    : enumType.GetOrCreateCacheEntry<EnumInfo<TStorage>>();
+            }
+        }
+
+        internal sealed partial class EnumInfo<TStorage> : RuntimeType.IGenericCacheEntry<EnumInfo<TStorage>>
+        {
+            public static EnumInfo<TStorage> Create(RuntimeType type. bool getNames)
             {
                 TStorage[]? values = null;
                 string[]? names = null;
 
                 GetEnumValuesAndNames(
-                    new QCallTypeHandle(ref enumType),
+                    new QCallTypeHandle(ref type),
                     ObjectHandleOnStack.Create(ref values),
                     ObjectHandleOnStack.Create(ref names),
                     getNames ? Interop.BOOL.TRUE : Interop.BOOL.FALSE);
 
                 Debug.Assert(values!.GetType() == typeof(TStorage[]));
-                Debug.Assert(!getNames || names!.GetType() == typeof(string[]));
 
-                bool hasFlagsAttribute = enumType.IsDefined(typeof(FlagsAttribute), inherit: false);
+                bool hasFlagsAttribute = type.IsDefined(typeof(FlagsAttribute), inherit: false);
+
+                return new EnumInfo<TStorage>(hasFlagsAttribute, values, names!);
+            }
+
+            public static EnumInfo<TStorage> Create(RuntimeType type) => Create(type, getNames: false);
+            {
```

Comment:
This needs to be deleted to make it build (github did not allowed me to do that in the suggestion).